### PR TITLE
Thomasht86/evaluator

### PR DESCRIPTION
**Pull request by thomasht86**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

This pull request introduces a new `VespaEvaluator`-module. This is something I've been missing for some time, that allows for easily evaluating a Vespa application across multiple metrics. 

The interface is heavily inspired by https://www.sbert.net/docs/package_reference/sentence_transformer/evaluation.html#informationretrievalevaluator and calculates the same metrics and produce the same output format. 

### New module `VespaEvaluator`

```python
from vespa.application import Vespa
from vespa.evaluation import VespaEvaluator

queries = {
    "q1": "What is the best GPU for gaming?",
    "q2": "How to bake sourdough bread?",
    # ...
}
relevant_docs = {
    "q1": {"d12", "d99"},
    "q2": {"d101"},
    # ...
}
# relevant_docs can also be a dict of query_id => single relevant doc_id
# relevant_docs = {
#     "q1": "d12",
#     "q2": "d101",
#     # ...
# }
def my_vespa_query_fn(query_text: str, top_k: int) -> dict:
    return {
        "yql": 'select * from sources * where userInput("' + query_text + '");',
        "hits": top_k,
        "ranking": "your_ranking_profile",
    }
app = Vespa(url="http://localhost", port=8080)
evaluator = VespaEvaluator(
    queries=queries,
    relevant_docs=relevant_docs,
    vespa_query_fn=my_vespa_query_fn,
    app=app,
    name="test-run",
    accuracy_at_k=[1, 3, 5],
    precision_recall_at_k=[1, 3, 5],
    mrr_at_k=[10],
    ndcg_at_k=[10],
    map_at_k=[100],
    write_csv=True
)
results = evaluator()
print("Primary metric:", evaluator.primary_metric)
print("All results:", results)
``` 
* Also added an example notebook, that compares 4 different ranking strategies `docs/sphinx/source/evaluating-vespa-application-cloud.ipynb` on [NanoMSMARCO](https://huggingface.co/datasets/zeta-alpha-ai/NanoMSMARCO) using the new class.

### New Tests:
* Added a new integration test for evaluating Vespa applications with `VespaEvaluator` in `tests/integration/test_integration_evaluation.py`. This test includes creating an application package, deploying it using VespaDocker, feeding documents, preparing queries and qrels, and comparing evaluation results to reference metrics from Sentence Transformers.
* Added a new unit test for `VespaEvaluator` in `tests/unit/test_evaluator.py`. This test covers various aspects such as initialization, custom k-values, accuracy, precision, recall, MRR, NDCG, MAP metrics, and validation of the Vespa query function.

### Documentation Updates:
* Updated the Sphinx documentation to include a new section on evaluating Vespa applications in the cloud (`evaluating-vespa-application-cloud`) in `docs/sphinx/source/index.rst`.
* Added documentation for the `VespaEvaluator` class in `docs/sphinx/source/reference-api.rst` under a new `vespa.evaluation` section.
* Documented the `schema_to_text` attribute of the `Schema` class in `docs/sphinx/source/reference-api.rst`.

### Workflow and Dependencies:
* Updated the GitHub workflow configuration in `.github/workflows/integration-except-cloud.yml` to include `test_integration_evaluation.py` in the list of integration tests.
* Added `matplotlib` to the list of dependencies for notebooks in `pyproject.toml`.

### CODE REVIEW

**Comment by thomasht86** [comment]
fyi 
@kkraune 
@ldalves

File: vespa/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+        # Validate the function can actually be called with test inputs
+        try:
+            result = fn("test query", 10)
+            if not isinstance(result, dict):
+                raise TypeError("vespa_query_fn must return a dict")
+        except Exception as e:
+            raise ValueError(f"Error calling vespa_query_fn with test inputs: {str(e)}")
+
+    def __call__(self) -> Dict[str, float]:
```
Comment (glebashnik) [inline vespa/evaluation.py:259]: run() method will be more explicit than making class callable IMHO.
Also doc strings missing for this one.

**Review by glebashnik — COMMENTED** [review]
This is great!
Just a small comment.

**Comment by glebashnik** [comment]
The notebook uses train part of the dataset.
This is fine for a quick test of unsupervised methods without much tweaking of the ranking profiles.
For more extensive experimentation, tweaking ranking profiles, parameters, fine-tuning embedders they will overfit and the results will not longer be representative.
Perhaps we should mention it.

**Review by glebashnik — CHANGES_REQUESTED** [review]
Looks great.
Some small comments.

File: vespa/evaluation.py (hunk 1 above)
Comment (thomasht86) [reply vespa/evaluation.py:259]: agree we should call run instead. will implement. also add docstring!

Comment (thomasht86) [reply vespa/evaluation.py]: should be good now! :)

**Review by glebashnik — APPROVED** [review]
